src/storage/src/sink/kafka.rs

```diff
@@ -257,6 +266,66 @@ struct KafkaTxProducer {
 }
 
 impl KafkaTxProducer {
+    async fn new<'a>(
+        KafkaTxProducerConfig {
+            name,
+            connection,
+            connection_context,
+            producer_context,
+            sink_id,
+            worker_id,
+        }: KafkaTxProducerConfig<'a>,
+    ) -> Result<KafkaTxProducer, ContextCreationError> {
+        let producer = connection
+            .create_with_context(
+                connection_context,
+                producer_context,
+                &btreemap! {
+                    // Ensure that messages are sinked in order and without
+                    // duplicates. Note that this only applies to a single
+                    // instance of a producer - in the case of restarts, all
+                    // bets are off and full exactly once support is required.
+                    "enable.idempotence" => "true".into(),
+                    // Increase limits for the Kafka producer's internal
+                    // buffering of messages Currently we don't have a great
+                    // backpressure mechanism to tell indexes or views to slow
+                    // down, so the only thing we can do with a message that we
+                    // can't immediately send is to put it in a buffer and
+                    // there's no point having buffers within the dataflow layer
+                    // and Kafka If the sink starts falling behind and the
+                    // buffers start consuming too much memory the best thing to
+                    // do is to drop the sink Sets the buffer size to be 16 GB
+                    // (note that this setting is in KB)
+                    "queue.buffering.max.kbytes" => format!("{}", 16 << 20),
+                    // Set the max messages buffered by the producer at any time
+                    // to 10MM which is the maximum allowed value.
+                    "queue.buffering.max.messages" => format!("{}", 10_000_000),
+                    // Make the Kafka producer wait at least 10 ms before
+                    // sending out MessageSets TODO(rkhaitan): experiment with
+                    // different settings for this value to see if it makes a
+                    // big difference.
+                    "queue.buffering.max.ms" => format!("{}", 10),
+                    "transactional.id" => format!("mz-producer-{sink_id}-{worker_id}"),
+                    // Time out transactions after 10 seconds
+                    "transaction.timeout.ms" => format!("{}", 10_000),
+                },
+            )
+            .await?;
+
+        let producer = KafkaTxProducer {
+            name,
+            inner: Arc::new(producer),
+            timeout: Duration::from_secs(5),
+        };
+
+        producer
+            .retry_on_txn_error(|p| p.init_transactions())
+            .await
+            .check_ssh_status(producer.inner.client().context())?;
```

Comment:
98% as safe and very pragmatic to leave all the code in the same place for now. Seems like the right call! 👍🏽